vochain/account.go

```diff
@@ -208,251 +223,248 @@ func (v *State) IncrementAccountProcessIndex(accountAddress common.Address) erro
-		return err
+// CreateAccount creates an account
+func (v *State) CreateAccount(accountAddress common.Address,
+	infoURI string,
+	delegates [][]byte) error {
+	return v.createAccount(accountAddress, infoURI, delegates, 0)
+}
+
+func (v *State) createAccount(accountAddress common.Address,
+	infoURI string,
+	delegates [][]byte,
+	initialBalance uint64,
+) error {
+	newAccount := &Account{}
+	if infoURI != "" && len(infoURI) <= types.MaxURLLength {
+		newAccount.InfoURI = infoURI
 	}
-	if err := v.SetAccount(BurnAddress, burnAcc); err != nil {
-		return err
+	if len(delegates) > 0 {
+		newAccount.DelegateAddrs = append(newAccount.DelegateAddrs, delegates...)
 	}
-	return nil
+	newAccount.Balance = initialBalance
+	log.Debugf("creating account %s with infoURI %s balance %d and delegates %+v",
+		accountAddress.String(),
+		newAccount.InfoURI,
+		newAccount.Balance,
+		printPrettierDelegates(newAccount.DelegateAddrs),
+	)
+	return v.SetAccount(accountAddress, newAccount)
 }
 
 // SetAccountInfoTxCheck is an abstraction of ABCI checkTx for an SetAccountInfoTx transaction
-// If the bool returned is true means that the account does not exist and is going to be created
-func SetAccountInfoTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (*setAccountInfoTxCheckValues, error) {
-	if vtx == nil {
-		return nil, ErrNilTx
-	}
-	returnValues := &setAccountInfoTxCheckValues{}
-	tx := vtx.GetSetAccountInfo()
-	// check signature available
-	if signature == nil || tx == nil || txBytes == nil {
-		return nil, fmt.Errorf("missing signature and/or transaction")
-	}
-	// check infoURI
-	infoURI := tx.GetInfoURI()
-	if infoURI == "" {
-		return nil, fmt.Errorf("invalid URI, cannot be empty")
+func SetAccountInfoTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) error {
+	if vtx == nil || signature == nil || txBytes == nil || state == nil {
+		return fmt.Errorf("invalid transaction parameters provided")
+	}
+	tx := vtx.GetSetAccount()
+	if tx == nil {
+		return fmt.Errorf("invalid transaction")
+	}
+	if tx.Nonce == nil || tx.InfoURI == nil {
+		return fmt.Errorf("missing signature and/or transaction")
 	}
-	// recover txSender address from signature
 	pubKey, err := ethereum.PubKeyFromSignature(txBytes, signature)
 	if err != nil {
-		return nil, fmt.Errorf("cannot extract public key from signature: %w", err)
+		return fmt.Errorf("cannot extract public key from signature: %w", err)
 	}
-	returnValues.TxSender, err = ethereum.AddrFromPublicKey(pubKey)
+	txSenderAddress, err := ethereum.AddrFromPublicKey(pubKey)
 	if err != nil {
-		return nil, fmt.Errorf("cannot extract address from public key: %w", err)
+		return fmt.Errorf("cannot extract address from public key: %w", err)
 	}
-	// get txSender account
-	txSender, err := state.GetAccount(returnValues.TxSender, false)
+	var txAccountAddress common.Address
+	if len(tx.Account) == 0 || bytes.Equal(tx.Account, types.EthereumZeroAddress.Bytes()) {
```

Comment:
Same with zero address... But in the other side I don't understand the reason behind this check. Why not just take the txSenderAddress and ignore Account? Or the other option would be to enforce tx.Account to be equal to txSenderAddress.

Reply:
because it can be a delegate and the signature points to the delegate address